regex/functions.go

```diff
@@ -0,0 +1,276 @@
+// Copyright 2023 Dolthub, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package regex
+
+import (
+	"context"
+	"fmt"
+)
+
+type URegularExpressionPtr uint32
+type UCharPtr uint32
+type UErrorCode int32
+type CharPtr int32
+
+type icuPointer interface {
+	URegularExpressionPtr | UCharPtr | CharPtr | uint32
+}
+
+// char* u_strToUTF8(char* dest, int32_t destCapacity, int32_t* pDestLength, const UChar* src, int32_t srcLength, UErrorCode* pErrorCode)
+func u_strToUTF8(ctx context.Context, buff CharPtr, bufflen int, outlen *int, str UCharPtr, strlen int, uerr *UErrorCode) (err error) {
+	strToUTF8 := module.ExportedFunction("u_strToUTF8_68")
+	var uerrptr uint32 = 0
+	if uerr != nil {
+		uerrptr, err = malloc[uint32](ctx, 4)
```

Comment:
I think these `malloc`s are overkill, if you want to do the leg work to get rid of them. emscripten put a statically sized region of memory which it treats as a stack in our WASM module. It copies local variables to it when it needs to pass a pointer to a local variable to a function. Technically we can do the same thing of course, and not go through malloc free list partitioning / etc. (I have no idea how malloc works in WASM, but it's got 11 local variables and is a large number of pages of disassembled code...so I'm guessing it's a real malloc...)

You can either use the exported functions `stackSave() i32`, `stackRestore(i32)`, `stackAlloc(i32) i32`, or you can grab the `module.ExportedGlobal("$global$0")` and use it as a mutable variable. AFAICT from the compiled code, the allocated stack offsets always need to be 16-aligned. So to use a UErrorCode, maybe something like:

```go
sp := module.ExportedGlobal("$global$0").(api.MutableGlobal)
origSP := sp.Get()
sp.Set(origSP - 16)
defer func() { sp.Set(origSP) }
uerraddr := origSP - 4
module.Memory().WriteUint32Le(uerrptr, uint32(*uerr))
defer func() {
    res, ok := module.Memory().ReadUint32Le(uerraddr)
    if !ok {
        err = fmt.Errorf("could not read UErrorCode")
    }
    *uerr = UErrorCode(res)
}()
```

This has a consequence of more tightly coupling us to emscripten in particular, since stack management in a WASM module is purely a complier convetion.